File: optimization/app/models/energy_signal.py
from datetime import datetime
from decimal import Decimal
from typing import Optional

from sqlalchemy import BigInteger, Column, DateTime, Enum, Numeric, String, UniqueConstraint
from sqlalchemy.dialects.postgresql import TIMESTAMP

from app.db import Base


class EnergySignal(Base):
    """
    Read-only mirror of the energy_signals table managed by the Spring backend.
    The optimization service only SELECTs from this table.
    """

    __tablename__ = "energy_signals"
    __table_args__ = (
        UniqueConstraint("signal_time", "source", name="uq_energy_signals_time_source"),
    )

    id: int = Column(BigInteger, primary_key=True, autoincrement=True)
    signal_time: datetime = Column(TIMESTAMP(timezone=True), nullable=False, index=True)
    source: str = Column(
        Enum("SOLAR", "WIND", "HYDRO", "GRID", "BATTERY", "MIXED", name="energy_source_enum"),
        nullable=False,
        default="MIXED",
    )
    renewable_pct: Decimal = Column(Numeric(5, 2), nullable=False, default=Decimal("0"))
    carbon_intensity_gco2_kwh: Optional[Decimal] = Column(Numeric(8, 2))
    electricity_price_per_kwh: Optional[Decimal] = Column(Numeric(8, 4))
    grid_load_pct: Optional[Decimal] = Column(Numeric(5, 2))
    solar_forecast_kw: Optional[Decimal] = Column(Numeric(10, 2))
    wind_forecast_kw: Optional[Decimal] = Column(Numeric(10, 2))
    fetched_at: datetime = Column(TIMESTAMP(timezone=True), nullable=False)

File: backend/src/main/java/com/evrenewable/model/EnergySignal.java
package com.evrenewable.model;

import lombok.*;

import javax.persistence.*;
import java.math.BigDecimal;
import java.time.Instant;

@Entity
@Table(name = "energy_signals",
       uniqueConstraints = @UniqueConstraint(columnNames = {"signal_time", "source"}))
@Getter @Setter @NoArgsConstructor @AllArgsConstructor @Builder
public class EnergySignal {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "signal_time", nullable = false)
    private Instant signalTime;

    @Enumerated(EnumType.STRING)
    @Column(name = "source", nullable = false, length = 20)
    private EnergySource source = EnergySource.MIXED;

    @Column(name = "renewable_pct", nullable = false, precision = 5, scale = 2)
    private BigDecimal renewablePct = BigDecimal.ZERO;

    @Column(name = "carbon_intensity_gco2_kwh", precision = 8, scale = 2)
    private BigDecimal carbonIntensityGco2Kwh;

    @Column(name = "electricity_price_per_kwh", precision = 8, scale = 4)
    private BigDecimal electricityPricePerKwh;

    @Column(name = "grid_load_pct", precision = 5, scale = 2)
    private BigDecimal gridLoadPct;

    @Column(name = "solar_forecast_kw", precision = 10, scale = 2)
    private BigDecimal solarForecastKw;

    @Column(name = "wind_forecast_kw", precision = 10, scale = 2)
    private BigDecimal windForecastKw;

    @Column(name = "fetched_at", nullable = false)
    private Instant fetchedAt = Instant.now();

    public enum EnergySource {
        SOLAR, WIND, HYDRO, GRID, BATTERY, MIXED
    }
}
